### PR TITLE
Added AVM Fritz!Powerline (#763)

### PR DESCRIPTION
**Pull request by cweitkamp**
- added support for AVM Fritz!Powerline

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
Um, ... headline for pull request ist wrong. Should be "Added AVM Fritz!Powerline (#763)"

**Comment by kaikreuzer** [comment]
Hm, all (most?) those new Powerline devices have the same functionality, do I see this right?
So is it needed to introduce different thing types for them or could it just be a single type that supports all of them in the very same way?

**Comment by cweitkamp** [comment]
Yes, that's correct. All Powerline Devices provide the same functionality except one (Powerline 546E). I don't think the new things are needed. I'm new to openhab, so that was the easiest way to solve the request. I will work out a solution for that.

**Comment by kaikreuzer** [comment]
Ok, refactoring it will be a good chance to learn a bit more about the binding APIs ;-)
Should really only mean to introduce a single thing type for those devices, while you can have the discovery service still put the specific name as a label to it.

**Comment by cweitkamp** [comment]
@kaikreuzer: I updated the thing configuration to use only one single thing type for all powerline devices with the same attributes.

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml [hunk 1]
```diff
@@ -45,126 +45,12 @@
...
-		</config-description>
-    </thing-type>
-
-    <thing-type id="FRITZ_Powerline_1000E_Solo">
-		<supported-bridge-type-refs>
-			<bridge-type-ref id="fritzbox" />
-		</supported-bridge-type-refs>
-        <label>FRITZ!Powerline 1000E</label>
-        <description>FRITZ!Powerline 1000E</description>
+        <label>FRITZ!Powerline devices</label>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml]: This is the label/desc that is shown for ONE device (that is added/discovered/etc.) - so I would suggest to change the label to "FRITZ!Powerline" and the description to "a FRITZ!Powerline device"

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java [hunk 2]
```diff
@@ -80,13 +80,16 @@ public DeviceHandler(Thing thing) {
      */
     @Override
     public void initialize() {
-        if (this.getThing().getThingTypeUID().equals(PL546E_STANDALONE_THING_TYPE)) {
-            logger.debug("About to initialize thing " + BindingConstants.DEVICE_PL546E_STANDALONE);
+        //if (this.getThing().getThingTypeUID().equals(PL546E_STANDALONE_THING_TYPE)) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java]: remove code in comment

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/i18n/avmfritz_de_DE.properties [hunk 3]
```diff
@@ -7,6 +7,14 @@ thing-type.Fritz_DECT_210.description = FRITZ!DECT210 schaltbare Steckdose
 thing-type.Fritz_DECT_200.description = Fritz!DECT200 schaltbare Steckdose
 thing-type.Fritz_DECT_Repeater_100.description = Fritz!DECT Repeater 100
 
+thing-type.FRITZ_Powerline_1240E.description = FRITZ!Powerline 1240E
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/i18n/avmfritz_de_DE.properties]: these thing types don't exist anymore, do they?

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -51,20 +45,74 @@
 		</config-description>
     </thing-type>
 
-    <thing-type id="FRITZ_Powerline_546E">
+    <thing-type id="Powerline_Solo">
 		<supported-bridge-type-refs>
 			<bridge-type-ref id="fritzbox" />
 		</supported-bridge-type-refs>
+        <label>FRITZ!Powerline devices</label>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml]: remove "devices"

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -51,20 +45,74 @@
...
+        <description>FRITZ!Powerline devices: 1240E, 1220E, 1000E, 540E, 530E, 510E</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml]: please apply the formatter - there is a tabs/spaces mix

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml]: Change desc to "a FRITZ!Powerline device"

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -51,20 +45,74 @@
 		</config-description>
     </thing-type>
 
-    <thing-type id="FRITZ_Powerline_546E">
+    <thing-type id="Powerline_Solo">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/thing/thing-types.xml]: do we really need the "_Solo" suffix?

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java [hunk 7]
```diff
@@ -29,25 +29,44 @@
...
-    public static final String PL546E_MODEL_NAME = "FRITZ!Powerline";
+    public static final String POWERLINE_MODEL_NAME = "FRITZ!Powerline";
     public static final String THING_AIN = "ain";
 
     // List of main device types
     public static final String DEVICE_DECT210 = "FRITZ_DECT_210";
     public static final String DEVICE_DECT200 = "FRITZ_DECT_200";
     public static final String DEVICE_DECT100 = "FRITZ_DECT_Repeater_100";
     public static final String DEVICE_PL546E = "FRITZ_Powerline_546E";
+    public static final String DEVICE_PL1240E_STANDALONE = "FRITZ_Powerline_1240E_Solo";
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java]: are these constants used?

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java [hunk 8]
```diff
@@ -29,25 +29,44 @@
...
+    
     // List of all Thing Type UIDs
     public final static ThingTypeUID BRIDGE_THING_TYPE = new ThingTypeUID(BINDING_ID, BRIDGE_FRITZBOX);
+    public final static ThingTypeUID POWERLINE_THING_TYPE = new ThingTypeUID(BINDING_ID, "Powerline_Solo");
 
     public final static ThingTypeUID DECT210_THING_TYPE = new ThingTypeUID(BINDING_ID, DEVICE_DECT210);
     public final static ThingTypeUID DECT200_THING_TYPE = new ThingTypeUID(BINDING_ID, DEVICE_DECT200);
     public final static ThingTypeUID DECT100_THING_TYPE = new ThingTypeUID(BINDING_ID, DEVICE_DECT100);
     public final static ThingTypeUID PL546E_THING_TYPE = new ThingTypeUID(BINDING_ID, DEVICE_PL546E);
+    public final static ThingTypeUID PL1240E_STANDALONE_THING_TYPE = new ThingTypeUID(BINDING_ID,
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java]: these thing types do not exist

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/HandlerFactory.java [hunk 9]
```diff
@@ -52,6 +52,12 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
         return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
     }
     /**
+     *
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/HandlerFactory.java]: don't add empty javadoc

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java (hunk 7 above)
Comment (cweitkamp) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java]: Yes, they are used to define the supported thing types later in this file.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java (hunk 8 above)
Comment (cweitkamp) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/BindingConstants.java]: Correct, but these constants are used to define the supported thing types later in this file.

If I understood the programming correctly the binding uses a xml-attribute "productname" of the interface to identify a supported device during discovery. This attribute contains the real description of the FRITZ! product (whitespaces are replaced by underscores).

**Pull request closed without merging**